### PR TITLE
feat(signals): allow defining private SignalStore members

### LINKED ISSUE #4443: RFC: Allow defining private SignalStore members
### Which @ngrx/* package(s) are relevant/related to the feature request?

signals

### Information

Provide the ability to define private SignalStore members by using the `_` prefix:

```ts
export const CounterStore = signalStore(
  withState({
    count1: 0,
    // 👇 private state slice
    _count2: 0,
  }),
  withComputed(({ count1, _count2 }) => ({
    // 👇 private computed signal
    _doubleCount1: computed(() => count1() * 2),
    doubleCount2: computed(() => _count2() * 2),
  })),
  withMethods((store) => ({
    increment1(): void {
      patchState(store, { count1: store.count1() + 1 });
    },
    // 👇 private method
    _increment2(): void {
      patchState(store, { _count2: store._count2() + 1 });
    },
  })),
);
```

All state slices, computed signals, and methods with the `_` prefix are not accessible from the outside:

```ts
const store = inject(CounterStore);

store.count1(); // ✅
store._count2(); // ❌

store._doubleCount1(); // ❌
store.doubleCount2(); // ✅

store.increment1(); // ✅
store._increment2(); // ❌
```

**Note:** Only root-level state slices can be defined as private.

Similar to private class fields in TypeScript, private SignalStore members will be excluded from the public store instance only at the type level:

```ts
(store as any)._increment2(); // works
```

## Updating private state slices when the state is unprotected

Even if the state is unprotected from external modifications, compilation error will be thrown if private state slices are updated from the outside:

```ts
const CounterStore = signalStore(
  { protectedState: false },
  withState({ count1: 0, _count2: 0 }),
);

const store = inject(CounterStore);
patchState(store, { _count2: 10 }); // ❌ compilation error
```

## Private members in custom SignalStore features

If a private member is defined in a custom SignalStore feature, it will be accessible in the store that uses it:

```ts
const withFoo = () => signalStoreFeature(withState({ _foo: 'bar' }));

const Store = signalStore(
  withFoo(),
  withComputed(({ _foo }) => ({
    bar: computed(() => _foo() + 'bar'),
  }))
);
```

To limit the visibility of custom store feature members, it's necessary to specify the return type explicitly:

```ts
export function withCount(): SignalStoreFeature<
  EmptyFeatureResult,
  { state: { count2: number }; computed: {}; methods: {} }
> {
  return signalStoreFeature(withState({ count1: 0, count2: 0 }));
}
```

The `count1` property won't be visible in the store that uses the `withCount` feature.

Another option is to use symbols as keys:

```ts
const FOO = Symbol('FOO');

export function withFoo() {
  return signalStoreFeature(
    withMethods({ [FOO]: () => {} })
  );
}
```

The `FOO` method will be visible in the store that uses the `withFoo` custom feature but it won't be accessible if the `FOO` symbol is not exported.

## Private entity collections

The `withEntities` feature will not allow specifying which entity properties are private or not. However, it will be possible to define all entity properties as private using a collection name with the `_` prefix. Then, public entity properties (state slices and computed signals) can be reexported with custom names:

```ts
const todoConfig = entityConfig({
  entity: type<Todo>(),
  // 👇 private collection
  collection: '_todo',
});

const TodosStore = signalStore(
  withEntities(todoConfig),
  withComputed(({ _todoEntities }) => ({
    // 👇 exposing entity array publicly
    todos: _todoEntities,
  }))
);

@Component({
  /* ... */
  template: `
    <h1>Todos</h1>
    <ngrx-todo-list [todos]="store.todos()" />
  `,
  providers: [TodosStore],
})
class TodosComponent {
  readonly store = inject(TodosStore);
}
```

### Describe any alternatives/workarounds you're currently using

_No response_

### I would be willing to submit a PR to fix this issue

- [X] Yes
- [ ] No

### PR DESCRIPTION
**Pull request by markostanimirovic**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngrx/platform/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Documentation has been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

Closes #4443

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: projects/ngrx.io/content/guide/signals/signal-store/entity-management.md [hunk 1]
```diff
@@ -384,3 +384,41 @@ export const TodosStore = signalStore(
 );
 
 </code-example>
+
+## Private Entity Collections
+
+Private entity collection can be defined by using the `_` prefix for the collection name.
```
Comment (brandonroberts) [inline projects/ngrx.io/content/guide/signals/signal-store/entity-management.md:390]: nit: phrasing

```suggestion
Private entity collections are defined by using the `_` prefix for the collection name.
```

File: projects/ngrx.io/content/guide/signals/signal-store/entity-management.md [hunk 2]
```diff
@@ -384,3 +384,41 @@ export const TodosStore = signalStore(
...
+  providers: [TodosStore],
+})
+class TodosComponent {
+  readonly store = inject(TodosStore);
+}
+```
+
+<div class="alert is-helpful">
+
+Learn more about private store members on [the following page](/guide/signals/signal-store/private-store-members).
```
Comment (brandonroberts) [inline projects/ngrx.io/content/guide/signals/signal-store/entity-management.md:422]:
```suggestion
Learn more about private store members in the [private store members guide](/guide/signals/signal-store/private-store-members).
```

File: projects/ngrx.io/content/guide/signals/signal-store/private-store-members.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
+# Private Store Members
+
+SignalStore allows defining private members that cannot be accessed from outside the store by using the `_` prefix.
```
Comment (brandonroberts) [inline projects/ngrx.io/content/guide/signals/signal-store/private-store-members.md:3]: Should we mention a compilation error is displayed when trying to access them?

**Review by brandonroberts — CHANGES_REQUESTED** [review]

File: projects/ngrx.io/content/guide/signals/signal-store/private-store-members.md (hunk 3 above)
Comment (markostanimirovic) [reply projects/ngrx.io/content/guide/signals/signal-store/private-store-members.md:3]: This was shown in the example: https://github.com/ngrx/platform/pull/4447/files#diff-9dd7a2eeb44718a61b7b0b462918eded72e902854e254161eb35629869d6f8f0R56

**Review by brandonroberts — APPROVED** [review]
LGTM

File: projects/ngrx.io/content/guide/signals/signal-store/private-store-members.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+export const CounterStore = signalStore(
+  withState({
+    count1: 0,
+    // 👇 private state slice
+    _count2: 0,
+  }),
+  withComputed(({ count1, _count2 }) => ({
+    // 👇 private computed signal
+    _doubleCount1: computed(() => count1() * 2),
```
Comment (michael-small) [inline projects/ngrx.io/content/guide/signals/signal-store/private-store-members.md:26]: nit: (applies to the test file as well) - the other two examples are a non-private and private example, respectively. However,  the `double` examples are private and then non-private. I think for consistency, In my opinion, I would change these two files to be `doubleCount` and `_doubleCount2` respectively.

Comment (markostanimirovic) [reply projects/ngrx.io/content/guide/signals/signal-store/private-store-members.md:26]: hey Michael,

This was intentional. I wanted to show that private members can be used to create public ones and vice versa.

Comment (michael-small) [reply projects/ngrx.io/content/guide/signals/signal-store/private-store-members.md:26]: That makes sense, sounds good. Thank you.

File: projects/ngrx.io/content/guide/signals/signal-store/entity-management.md [hunk 5]
```diff
@@ -384,3 +384,41 @@ export const TodosStore = signalStore(
...
+
+## Private Entity Collections
+
+Private entity collections are defined by using the `_` prefix for the collection name.
+
+```ts
+const todoConfig = entityConfig({
+  entity: type<Todo>(),
+  // 👇 private collection
+  collection: '_todo',
```
Comment (timdeschryver) [inline projects/ngrx.io/content/guide/signals/signal-store/entity-management.md:396]: Should we add a JSDoc comments for the `collection` property to mention this behavior?
I'm thinking that new devs wouldn't expect this behavior by naming a collection.

**Review by timdeschryver — APPROVED** [review]

File: projects/ngrx.io/content/guide/signals/signal-store/entity-management.md (hunk 5 above)
Comment (markostanimirovic) [reply projects/ngrx.io/content/guide/signals/signal-store/entity-management.md:396]: Definitely! I'm going to add JSDoc for all public APIs.

**Pull request merged**